### ISSUE TITLE
Re-use BugReporting.jl

### ISSUE THREAD

**Issue by maleadt** [JuliaCI/julia-buildkite#186]
Our CI set-up here has its own utilities for working with rr traces: https://github.com/JuliaCI/julia-buildkite/tree/main/utilities/rr
I wonder if it wouldn't be better to re-use BugReporting.jl, and just launch Julia under `--bug-report=rr`. Much of the needed functionality (packing and uploading traces, handling time-outs, doing replay...) is already part of BugReporting.jl, and there's some useful features that aren't available here (packaging additional metadata, downloading and importing Julia sources on replay, ...).

**Comment by DilumAluthge** [comment]
A few thoughts:

1. IIRC, the `rr_capture.jl` script in this repo is designed to correctly capture traces for all of the Distributed worker processes that we launch as part of the base Julia test suite. Does BugReporting.jl also do that?
2. BugReporting.jl uploads to S3 but requires a manual interactive login. Obviously that doesn't work for us. But also, our test jobs are untrusted, so we cannot include any secrets (GitHub tokens, S3 keys, etc.) in the test job. So we need to add a feature to BugReporting.jl that detects if we are running on Buildkite, and if we are, uploads the traces as Buildkite artifacts instead of uploading to S3. Pseudo-code for doing so:

```julia
const list_of_files_to_upload = ... # get this list from the appropriate place

const is_buildkite = get(ENV, "BUILDKITE", "") == "true"
const is_ci = get(ENV, "BUILDKITE", "CI") == "true"
if is_buildkite && is_ci
    for file in list_of_files_to_upload
        run(`buildkite-agent artifact upload $(file)`)
    end
else
    do_upload_to_s3_with_manual_interactive_github_authentication(list_of_files_to_upload) 
end
```

**Comment by maleadt** [comment]
No, BugReporting.jl doesn't have those features yet, that's why I opened this issue. I'm happy to look into adding these features though, because I think it would be pretty beneficial if we didn't duplicate this functionality like we do now.
